### PR TITLE
feat(extension): LW-6779 on-demand addresses discovery

### PR DESCRIPTION
**Pull request by szymonmaslowski**
# Checklist

- [x] JIRA - [LW-6779](https://input-output.atlassian.net/browse/LW-6779), [W-8841](https://input-output.atlassian.net/browse/LW-8841), [LW-8842](https://input-output.atlassian.net/browse/LW-8842), [LW-8843](https://input-output.atlassian.net/browse/LW-8843)
- [ ] Proper tests implemented
- [x] Screenshots added.

---

## Proposed solution

**A button in settings triggering a address discovery**

At the current point in time, there is no simple way to use the SDK API to trigger the HD wallet discovery. Therefore a hacky way has been developed that achieves the following goals:
 - allow for triggering the discovery on demand
 - catch the discovery status initiated by the PersonalWallet or on-demand request
 - catching any errors happening during the discovery

Implementation details:
There is an `addressesDiscoverer` logic living in the service worker. It consists of:
 - the `addressesDiscoverer` public interface accessible via remote API for the frontend instances. It exposes
   - `setup` method for configuring the `addressesDiscoverer` for specific network
   - `discover` method for triggering the discovery on-demand
   - `status$` stream providing the current discovery status
 - the `getDiscovererInstance` remembering and returning an instance of a discovery logic (wrapped `HDSequentialDiscovery` object) for a given network. That instance gets accessed during the creation of the `PersonalWallet` instance and during the on-demand discovery request using the `addressesDiscoverer` API.

Every time the network changes the `addressesDiscoverer` gets updated with the new `chainName` and the new asyncKeyAgent which is being exposed to the `addressesDiscoverer` through the remote API.

## Testing

Scenario 1:
1. Install extension on two browsers/accounts
2. restore the same wallet in both places
4. make a delegation in the first instance to at least one more pool and wait till it finishes
6. go to staking in the second instance and confirm there is one pool missing
7. go to settings
8. click sync.
9. confirm you have all the pools

Scenario 2:
1. Restore a wallet and click go to my wallet button fast enough
2. see the sync overlay

## Screenshots

Settings option:
image
image
Overlay:
image
image
Success toast:
image
image
Error toast:
image
image

---
# Allure report
`allure-report-publisher` generated test report!

**smokeTests**: ❌ [test report](https://lace-qa:8PP5wBaDV6UoXj@dq4ajm5i5q7bz.cloudfront.net/linux/chrome/2725/6660158911/index.html) for [e2a0291d](https://github.com/input-output-hk/lace/pull/673/commits/e2a0291d2341a04da03e64fe94342f35981140b5)
|       | passed | failed | skipped | flaky | total | result |
|-------|--------|--------|---------|-------|-------|--------|
| Total | 3      | 29     | 0       | 0     | 32    | ❌     |

[LW-6779]: https://input-output.atlassian.net/browse/LW-6779?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[LW-8842]: https://input-output.atlassian.net/browse/LW-8842?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[LW-8843]: https://input-output.atlassian.net/browse/LW-8843?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/browser-extension-wallet/src/hooks/useAddressesDiscoverer.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { Shutdown } from '@cardano-sdk/util/dist/esm';
```
Comment (przemyslaw-wlodek) [inline apps/browser-extension-wallet/src/hooks/useAddressesDiscoverer.ts:1]: Please fix the import path. :)

File: apps/browser-extension-wallet/src/hooks/useAddressesDiscoverer.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    const { channelName, shutdown } = await exposeKeyAgent(asyncKeyAgent);
+    shutdownExposedKeyAgent = shutdown;
+    await addressesDiscoverer.setup({ chainName, keyAgentChannelName: channelName });
+  }, [])
+});
+
+// TODO: for now just to have a way to trigger it. Soon it will be triggered with a button in settings
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+window.addressesDiscoverer = addressesDiscoverer;
```
Comment (przemyslaw-wlodek) [inline apps/browser-extension-wallet/src/hooks/useAddressesDiscoverer.ts:28]: Can we remove that? It won't be needed as soon as we merge it to the integration branch with the other changes.

File: apps/browser-extension-wallet/src/lib/translations/en.json [hunk 3]
```diff
@@ -1296,5 +1301,12 @@
       "usedAddresses.addressCopied": "Address copied",
       "showUsedAddresses": "Show used addresses"
     }
+  },
+  "addressesDiscoveryOverlay": {
+    "overlay.title": "Your wallet is syncing, this might take a few minutes",
```
Comment (przemyslaw-wlodek) [inline apps/browser-extension-wallet/src/lib/translations/en.json:1306]: Nit: "title" is enough, now it duplicates the parent key.

File: apps/browser-extension-wallet/src/views/browser-view/components/WarningModal/WarningModal.tsx [hunk 4]
```diff
@@ -60,9 +62,16 @@ export const WarningModal = ({
             {cancelLabel ?? translate('general.button.cancel')}
           </Button>
         )}
-        <Button className={confirmCustomClassName} data-testid="delete-address-modal-confirm" block onClick={onConfirm}>
-          {confirmLabel ?? translate('general.button.agree')}
-        </Button>
+        {withConfirm && (
```
Comment (przemyslaw-wlodek) [inline apps/browser-extension-wallet/src/views/browser-view/components/WarningModal/WarningModal.tsx:65]: Why do we need this? I assume we should base this logic on `onConfirm` and `onCancel`. If callbacks aren't available then buttons aren't rendered. `withCancel` has only 1 usage in Lace.

Maybe it's not for this PR but adding `withConfirm` is following the wrong path.

**Review by przemyslaw-wlodek — CHANGES_REQUESTED** [review]

**Review by emiride — APPROVED** [review]
Works as expected

**Pull request merged**